### PR TITLE
refactor(sns): migrate sns governance to use ic_cdk

### PR DESCRIPTION
**Pull request by max-dfinity**
This migrates sns governance to use ic_cdk instead of dfn_core

### CODE REVIEW

File: rs/sns/governance/canister/canister.rs [hunk 1]
```diff
@@ -104,10 +103,7 @@ impl CanisterEnv {
             // unpredictability since the pseudo-random numbers could still be predicted after
             // inception.
             rng: {
-                let now_nanos = now()
-                    .duration_since(SystemTime::UNIX_EPOCH)
-                    .unwrap()
-                    .as_nanos();
+                let now_nanos = now_nanoseconds() as u128;
```
Comment (aterga) [inline rs/sns/governance/canister/canister.rs:106]: NCR if you decide this is okay.

File: rs/sns/governance/canister/canister.rs [hunk 2]
```diff
@@ -702,52 +587,34 @@ fn encode_metrics(w: &mut ic_metrics_encoder::MetricsEncoder<Vec<u8>>) -> std::i
...
-    over(candid, |_: ()| include_str!("governance.did").to_string())
-}
-#[cfg(feature = "test")]
-#[export_name = "canister_query __get_candid_interface_tmp_hack"]
-fn expose_candid() {
-    over(candid, |_: ()| {
-        include_str!("governance_test.did").to_string()
-    })
+#[query(hidden = true)]
+fn __get_candid_interface_tmp_hack() -> String {
```
Comment (aterga) [inline rs/sns/governance/canister/canister.rs:591]: Do we still need this?

File: rs/sns/governance/canister/canister.rs [hunk 3]
```diff
@@ -702,52 +587,34 @@ fn encode_metrics(w: &mut ic_metrics_encoder::MetricsEncoder<Vec<u8>>) -> std::i
...
-#[export_name = "canister_query get_upgrade_journal"]
-fn get_upgrade_journal() {
-    over(candid_one, get_upgrade_journal_)
-}
-
-#[candid_method(query, rename = "get_upgrade_journal")]
-fn get_upgrade_journal_(arg: GetUpgradeJournalRequest) -> GetUpgradeJournalResponse {
+#[query]
+fn get_upgrade_journal(arg: GetUpgradeJournalRequest) -> GetUpgradeJournalResponse {
     let GetUpgradeJournalRequest {} = arg;
```
Comment (aterga) [inline rs/sns/governance/canister/canister.rs:610]: This is what I call top programming sophistication. :-)

NCR

File: rs/sns/governance/src/governance.rs [hunk 4]
```diff
@@ -73,10 +73,13 @@ use crate::{
     },
 };
 use candid::{Decode, Encode};
-use dfn_core::api::{spawn, CanisterId};
-use ic_base_types::PrincipalId;
+#[cfg(not(target_arch = "wasm32"))]
+use futures::FutureExt;
```
Comment (aterga) [inline rs/sns/governance/src/governance.rs:77]: Why did this become needed?

File: rs/sns/governance/src/governance.rs [hunk 5]
```diff
@@ -699,6 +703,24 @@ pub struct Governance {
     pub test_features_enabled: bool,
 }
 
+/// This function is used to spawn a future in a way that is compatible with both the WASM and
+/// non-WASM environments that are used for testing.  This only actually spawns in the case where
+/// the WASM is running in the IC, or has some other source of asynchrony.  Otherwise, it
+/// immediately executes.s
```
Comment (aterga) [inline rs/sns/governance/src/governance.rs:709]:
```suggestion
/// immediately executes.
```

File: rs/sns/governance/src/governance.rs [hunk 6]
```diff
@@ -4602,7 +4621,7 @@ impl Governance {
 
     /// Runs periodic tasks that are not directly triggered by user input.
     pub async fn heartbeat(&mut self) {
-        use dfn_core::println;
+        use ic_cdk::println;
```
Comment (aterga) [inline rs/sns/governance/src/governance.rs:4624]: Why don't we import this at the top of the file?

File: rs/sns/governance/src/types.rs [hunk 7]
```diff
@@ -2014,6 +2014,13 @@ pub struct LedgerUpdateLock {
 impl Drop for LedgerUpdateLock {
     /// Drops the lock on the neuron.
```
Comment (aterga) [inline rs/sns/governance/src/types.rs:2015]: Is this comment misleading? 

```suggestion
```

File: rs/sns/governance/src/types.rs [hunk 8]
```diff
@@ -2014,6 +2014,13 @@ pub struct LedgerUpdateLock {
...
     fn drop(&mut self) {
+        // In the case of a panic, the state of the ledger account representing the neuron's stake
+        // may be inconsistent with the internal state of governance.  In that case,
+        // we want to prevent further operations with that neuron until the issue can be
+        // investigated and resolved, which will require code changes.
+        if ic_cdk::api::call::is_recovering_from_trap() {
```
Comment (aterga) [inline rs/sns/governance/src/types.rs:2021]: Let's discuss this over zoom

File: rs/sns/governance/canister/canister.rs [hunk 9]
```diff
@@ -154,7 +145,9 @@ impl Environment for CanisterEnv {
             /* message: */ String,
         ),
     > {
-        call_bytes_with_cleanup(canister_id, method_name, &arg, Funds::zero()).await
+        ic_cdk::api::call::call_raw(canister_id.get().0, method_name, &arg, 0)
```
Comment (aterga) [inline rs/sns/governance/canister/canister.rs:148]: Could you add a comment explaining why we're not using just `ic_cdk::call` here?

File: rs/sns/governance/canister/canister.rs [hunk 10]
```diff
@@ -176,23 +169,39 @@ impl Environment for CanisterEnv {
...
     /// Return the canister version.
     fn canister_version(&self) -> Option<u64> {
-        Some(dfn_core::api::canister_version())
+        Some(ic_cdk::api::canister_version())
+    }
+}
+
+fn now_nanoseconds() -> u64 {
+    if cfg!(target_arch = "wasm32") {
+        ic_cdk::api::time()
```
Comment (aterga) [inline rs/sns/governance/canister/canister.rs:183]: NCR; 

I'm again wondering why this semantics isn't built into `ic_cdk::api::time` ...

**Review by aterga — APPROVED** [review]
Pre-approving, but I'd like to know more about the Drop implementation before we merge.

File: rs/sns/governance/canister/canister.rs (hunk 2 above)
Comment (max-dfinity) [reply rs/sns/governance/canister/canister.rs:591]: Not sure.  Would probably remove this in an unrelated PR.

File: rs/sns/governance/canister/canister.rs (hunk 9 above)
Comment (max-dfinity) [reply rs/sns/governance/canister/canister.rs:150]: done

**Pull request merged**